### PR TITLE
zypper_repository: Fix repo parsing for empty list

### PR DESCRIPTION
**Pull request by toabctl**
When no repositories are defined in zypper, the return code
of "zypper repos" is 6. Handle that case and don't fail
if zypper_repository has to deal with an empty repo list.

Fixes https://github.com/ansible/ansible-modules-extras/issues/795

### CODE REVIEW

File: packaging/os/zypper_repository.py [hunk 1]
```diff
@@ -95,21 +95,29 @@ def zypper_version(module):
...
+        return repos
+    # exit code 6 is ZYPPER_EXIT_NO_REPOS (no repositories defined)
+    elif rc == 6:
+        return []
+    else:
+        #module.fail_json(msg=stderr if stderr else stdout)
+        if stderr:
+            module.fail_json(msg=stderr)
+        else:
+            module.fail_json(msg=stdout)
```
Comment (abadger) [inline packaging/os/zypper_repository.py]: The old code didn't fail if it couldn't parse the repo files.  I don't use zypper but I think is reasonable to fail in this case.  Probably want to include the rc code as a field of the return value.

I also don't know what the stdout and stderr might contain.  If they aren't really detaialed might be good to include some information about where this is failing.  Something like:

``` python

msg='Failed to parse existing zypper repository files: %s' % stderr
```

**Comment by abadger** [comment]
Just one comment about including more information if we fail in this step.  Once that's fixed, this is good to merge.

**Comment by toabctl** [comment]
> Just one comment about including more information if we fail in this step. Once that's fixed, this is good to merge.

@abadger Done in https://github.com/ansible/ansible-modules-extras/commit/7d9988c96119ce03525d792f7d4d3e8fd9ecfe22 . But I wonder if I should just say in the `msg` the exact command and add `stderr` and `stdout`  to the json as extra keys. Something like:

```
d = { 'zypper_exit_code': rc }
if stderr:
    d['stderr'] = stderr
if stdout:
    d['stdout'] = stdout                                                                                                                                            
module.fail_json(msg='Failed to execute "%s"' % " ".join(cmd), **d)
```

But `stdout` and `stderr` are anyway printed in an error case, right? So that's just useless overhead I guess.
No idea what the common pattern in ansible is. Or is this good enough as it is now?

**Comment by abadger** [comment]
Your idea sounds good.  stdout and stderr I believe are only printed if the
module exits with a non-zero exit code.  Returning via fail_json() doesn't
print stdout and stderr (this lets us handle errors more gracefully as we
can control precisely what goes into the fail_json() data.

**Comment by toabctl** [comment]
@abadger Done now. Fine to merge it or still something todo?

**Comment by abadger** [comment]
Looks good.  Thanks!

# 

Hi!

This has been merged in, and will also be included in the next major release.

If you or anyone else has any further questions, please let us know by stopping by one of the two mailing lists, as appropriate:
- https://groups.google.com/forum/#!forum/ansible-project - for user questions, tips, and tricks
- https://groups.google.com/forum/#!forum/ansible-devel - for strategy, future planning, and questions about writing code

Because this project is very active, we're unlikely to see comments made on closed tickets, but the mailing list is a great way to ask questions, or post if you don't think this particular
issue is resolved.

Thank you!

**Pull request merged**